migration_scripts/bulk_load/src/footprint.clj

```diff
@@ -0,0 +1,263 @@
+           FROM
+               pg_class
+           WHERE
+               relkind IN ('i', 'r')
+           AND relname NOT LIKE 'pg_%'
+           AND relname NOT LIKE 'sql_%'
+           AND relname NOT LIKE '__diesel_%'"]
+         (jdbc/execute! db)
+
+         ;; Each row represents a table or an index. Figure out
+         ;; which (potentially partitioned) table it belongs to.
+         (map (fn [{:pg_class/keys [relname relkind reltuples]
+                    :keys [tab_size rel_size]}]
+                [(table->entity relname)
+                 (cond
+                   (= relkind "r")
+                   {:self rel_size
+                    :toast (- tab_size rel_size)
+                    :tuples reltuples}
+                   (s/ends-with? relname "_pkey")
+                   {:pkey rel_size}
+                   :else
+                   {:idx rel_size})]))
+
+         ;; Group by the table name, and merge entries together, so that for
+         ;; each table we get a single stats map.
+         (group-by first)
+         (map (fn [[key stats]]
+                [key (apply merge-with +
+                            (map second stats))]))
+
+         ;; Sort results so that partitions end up in order, and group again,
+         ;; this time by table names ignoring partitions.
+         (sort-by first)
+         (group-by (comp first first))
+         (map (fn [[key stats]]
+                [key (->> stats (map second) (into []))]))
+
+         ;; Gather results into a dictionary keyed by table names, ignoring
+         ;; partitions.
+         (into {}))))
+
+(defn pretty-size
+  "Pretty print byte sizes."
+  [sz]
+  (loop [sz sz units '("B" "KiB" "MiB" "GiB" "TiB")]
+    (if (or (< sz 1024) (not (rest units)))
+      (format "%.2f %s" (double sz) (first units))
+      (recur (/ sz 1024) (rest units)))))
+
+(defn footprint
+  "Calculate the total footprint given the per-table breakdown."
+  [tables]
+  (->> tables
+     (mapcat (fn [[_ stats]] stats))
+     (map (fn [{:keys [pkey self idx toast]}]
+            (+ self (or pkey 0) (or idx 0) (or toast 0))))
+     (reduce +)))
+
+(defn partitions
+  "The number of daily partitions in the table stats.
+
+  Fails if different partitioned tables disagree on how many
+  partitions there should be."
+  [tables]
+  (let [objects-history (-> tables :objects-history count)
+        transactions    (-> tables :transactions    count)
+        events          (-> tables :events          count)]
+    (assert (= objects-history transactions events))
+    objects-history))
+
+;; Clustered Tables ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn clustered
+  "Simulate storing the data in an RDBMS that supports clustered tables.
+
+  In that case, we will use no extra storage for primary keys"
+  [tables]
+  (let [no-pkey #(dissoc % :pkey)]
+    (->> tables
+         (map (fn [[table stats]]
+                [table (->> stats (map no-pkey) (into []))]))
+         (into {}))))
```

Comment:
what is this in preparation for?

Reply:
This is one of the entrypoint functions (This file is used from the REPL or a scratch buffer where you can load in statistics and then call this and other functions on them to see how the stats change.